### PR TITLE
Efficient swa

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
Swa recipe seems to scale quite well but takes very long to train somehow it has mediocre results until some critical point.

Trying to create a swa recipe with much fewer total epochs

### CODE REVIEW

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Hmm it seems that longer training stages (especially at low lr) seems to increase the sparsity but also the loss.

To really unlock longer training stages it's probably necessary to introduce some sort of activation loss as regularization to prevent dead regions.


Might also be an issue with optimizer state + weight clipping. The momentum might explode for some weights

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Oh.. stage 2 wasn't finished yet... Now might as well update trainer for stage 2 as well...

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Ah... New trainer sha is not compatible with older checkpoint... I'd wait for efficient swa 2 to finish first stage so that the training isn't redundant...

**Comment by TonyCongqianWang** [comment]
Stage 1 is almost done at #320

**Comment by TonyCongqianWang** [comment]
Stage 1 is now done :)

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Last stage didn't seem to achieve anything so I tuned it slightly..

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
build got kiled for whatever reason...

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Also difficult to tell given the error bars... i will remove sprt from testing to enable tuning

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
I hope the accuracy is now good enough to see a difference...

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Not super conclusive.. perhaps the exact lambda isn't that important..

**Comment by TonyCongqianWang** [comment]
Trying  a run that matches current master in time budget

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Not bad for a first full run but is being outperformed by the new skip logic alone. So needs some tuning probably

**Comment by TonyCongqianWang** [comment]
Testing all to get some data on the progress to decide for the new hyperparams...

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Using an entirely new, multi-cycle recipe with newest sha

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Very very good result not even considering the ~4000 epoch budget

**Comment by TonyCongqianWang** [comment]
I wonder if the warmup stage is really still necessary

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
I merged main into feature branch and added the epoch budget from the removed warmup stage to the first real stage for better comparison

**Comment by vondele** [comment]
cscs-ci run RECIPE=threats

**Comment by TonyCongqianWang** [comment]
Seems quite a bit worse without the special warmup stage... So guess It has to be left in for now..
